rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala

```diff
@@ -0,0 +1,151 @@
+import coop.rchain.rspace.internal.{Datum, Row}
+import monix.eval.{Coeval, Task}
+import monix.execution.Scheduler.Implicits.global
+import org.scalactic.TripleEqualsSupport
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.duration._
+
+class CostAccountingReducerTest extends FlatSpec with Matchers with TripleEqualsSupport {
+
+  behavior of "Cost accounting in Reducer"
+
+  it should "charge for the successful substitution" in {
+    val term: Expr => Par = expr => Par(bundles = Seq(Bundle(Par(exprs = Seq(expr)))))
+    val varTerm           = term(Expr(EVarBody(EVar(Var(BoundVar(0))))))
+    val substTerm         = term(Expr(GString("1")))
+    val env               = Env.makeEnv[Par](Expr(GString("1")))
+    val termCost          = Chargeable[Par].cost(substTerm)
+    val initCost          = CostAccount(1000)
+    val costAlg           = CostAccountingAlg.unsafe[Coeval](initCost)
+    val res               = Reduce.substituteAndCharge[Par, Coeval](varTerm, 0, env, costAlg).attempt.value
+    assert(res === Right(substTerm))
+    assert(costAlg.get().value.cost === (initCost.cost - Cost(termCost)))
+  }
+
+  it should "charge for failed substitution" in {
+    val term: Expr => Par = expr => Par(bundles = Seq(Bundle(Par(exprs = Seq(expr)))))
+    val varTerm           = term(Expr(EVarBody(EVar(Var(FreeVar(0))))))
+    val env               = Env.makeEnv[Par](Expr(GString("1")))
+    val originalTermCost  = Chargeable[Par].cost(varTerm)
+    val initCost          = CostAccount(1000)
+    val costAlg           = CostAccountingAlg.unsafe[Coeval](initCost)
+    val res               = Reduce.substituteAndCharge[Par, Coeval](varTerm, 0, env, costAlg).attempt.value
+    assert(res.isLeft)
+    assert(costAlg.get().value.cost === (initCost.cost - Cost(originalTermCost)))
+  }
+
+  it should "stop if OutOfPhloError is returned from RSpace" in {
+    val tuplespaceAlg = new TuplespaceAlg[Task] {
+      override def produce(
+          chan: Channel,
+          data: ListChannelWithRandom,
+          persistent: Boolean
+      ): Task[Unit] =
+        Task.raiseError(OutOfPhlogistonsError)
+      override def consume(
+          binds: Seq[(BindPattern, ChannelInstance.Quote)],
+          body: ParWithRandom,
+          persistent: Boolean
+      ): Task[Unit] = Task.raiseError(OutOfPhlogistonsError)
+    }
+
+    implicit val errorLog = new ErrorLog()
+    implicit val rand     = Blake2b512Random(128)
+    implicit val costAlg  = CostAccountingAlg.unsafe[Task](CostAccount(1000))
+    val reducer           = new DebruijnInterpreter[Task, Task.Par](tuplespaceAlg, Map.empty)
+    val send              = Send(Channel(Quote(GString("x"))), Seq(Par()))
+    val test              = reducer.inj(send).attempt.runSyncUnsafe(1.second)
+    assert(test === Left(OutOfPhlogistonsError))
+  }
+
+  it should "stop interpreter threads as soon as deploy runs out of phlo" in {
```

Comment:
We don't have the concept of "first" and "second"contract when they run in parallel so our test should not try to make that difference.

Instead what we should do is that we should execute two things in parallel and assert that the resulting cost is an error and the consumed phlos are either the cost of the first one or the cost of the second one but not both

Reply:
I can't see how to test that because we don't control how fast the threads reduce. This test is set up to have the _slower_ branch reduce later and have the phlos used before it hits the tuplespace.

Reply:
Although I think this test can be generalized so it doesn't care which expression reduced first and just check the final tuplespace content and cost state.

Reply:
if we have the code `p1 | p2` where p1 has cost c1 and p2 has cost c2 then we have two different valid outcomes for the resulting cost: c1 or c2. We don't care which because both are valid.

Reply:
aren't the costs the same now? I think you should use strings with different sizes such that the left and right side of | have different costs

Reply:
They are the same but it doesn't matter. The effect is that only one of the produces makes it to the tuplespace meaning that the other one was interrupted.